### PR TITLE
make use of new kernel 5.9 close_range() syscall in close_all_fds()

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by mrc0mmand** [comment]
This appears to cause all the havoc:

```
284/790 test-fdset                                                                                  FAIL           0.38s (exit status 1)

--- command ---
19:20:09 SYSTEMD_KBD_MODEL_MAP='/build/src/locale/kbd-model-map' SYSTEMD_LANGUAGE_FALLBACK_MAP='/build/src/locale/language-fallback-map' PATH='/systemd-meson-build:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl' /systemd-meson-build/asan-wrapper-GHz.sh /systemd-meson-build/test-fdset
--- stderr ---
=================================================================
==47604==ERROR: AddressSanitizer: dynamic-stack-buffer-overflow on address 0x7ffddc560164 at pc 0x7f0eb3a260e9 bp 0x7ffddc55ff70 sp 0x7ffddc55ff68
READ of size 4 at 0x7ffddc560164 thread T0
    #0 0x7f0eb3a260e8 in fd_in_set /systemd-meson-build/../build/src/basic/fd-util.c:187:21
    #1 0x7f0eb3a24ba4 in close_all_fds /systemd-meson-build/../build/src/basic/fd-util.c:351:21
    #2 0x7f0eb387e26b in fdset_close_others /systemd-meson-build/../build/src/shared/fdset.c:224:16
    #3 0x560bfe7822e7 in test_fdset_close_others /systemd-meson-build/../build/src/test/test-fdset.c:85:9
    #4 0x560bfe781387 in main /systemd-meson-build/../build/src/test/test-fdset.c:195:9
    #5 0x7f0eb317f001 in __libc_start_main (/usr/lib/libc.so.6+0x27001)
    #6 0x560bfe78129d in _start (/systemd-meson-build/test-fdset+0x429d)

Address 0x7ffddc560164 is located in stack of thread T0
SUMMARY: AddressSanitizer: dynamic-stack-buffer-overflow /systemd-meson-build/../build/src/basic/fd-util.c:187:21 in fd_in_set
Shadow bytes around the buggy address:
  0x10003b8a3fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a3fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a3ff0: 00 00 00 00 00 00 00 00 ca ca ca ca 00 cb cb cb
  0x10003b8a4000: cb cb cb cb 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a4010: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10003b8a4020: 00 00 00 00 00 00 00 00 ca ca ca ca[04]cb cb cb
  0x10003b8a4030: cb cb cb cb 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a4040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a4050: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a4060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10003b8a4070: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==47604==ABORTING
```

**Comment by mrc0mmand** [comment]
And I guess this build fail with `--optimization=2` is related:

```
In file included from /usr/include/string.h:494,
                 from ../src/basic/alloc-util.h:7,
                 from ../src/basic/fd-util.c:9:
In function ‘memcpy’,
    inlined from ‘close_all_fds’ at ../src/basic/fd-util.c:255:33:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: error: ‘__builtin_memcpy’ writing 18446744073709551612 bytes into a region of size 1 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/basic/fd-util.c:9:
../src/basic/fd-util.c: In function ‘close_all_fds’:
../src/basic/alloc-util.h:96:16: note: at offset 0 to an object with size 1 allocated by ‘malloc’ here
   96 |         return malloc(size * need ?: 1);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
[65/1704] Compiling C object src/basic/libbasic.a.p/fileio.c.o
[66/1704] Compiling C object src/basic/libbasic.a.p/fs-util.c.o
ninja: build stopped: subcommand failed.
+ fatal 'ninja failed with --optimization=2'
+ echo -e '\033[31;1mninja failed with --optimization=2\033[0m'
ninja failed with --optimization=2
```

**Comment by keszybz** [comment]
```
In file included from /usr/include/string.h:494,
                 from ../src/basic/alloc-util.h:7,
                 from ../src/basic/fd-util.c:9:
In function ‘memcpy’,
    inlined from ‘close_all_fds’ at ../src/basic/fd-util.c:256:33:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: error: ‘__builtin_memcpy’ writing 18446744073709551612 bytes into a region of size 1 [-Werror=stringop-overflow=]
   34 |   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/basic/fd-util.c:9:
../src/basic/fd-util.c: In function ‘close_all_fds’:
../src/basic/alloc-util.h:96:16: note: at offset 0 to an object with size 1 allocated by ‘malloc’ here
   96 |         return malloc(size * need ?: 1);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
Hmmmmm.

File: src/basic/alloc-util.h [hunk 1]
```diff
@@ -27,22 +27,22 @@ typedef void (*free_func_t)(void *p);
                 size_t _n_ = n;                                         \
                 assert(!size_multiply_overflow(sizeof(t), _n_));        \
                 assert(sizeof(t)*_n_ <= ALLOCA_MAX);                    \
-                (t*) alloca(sizeof(t)*_n_);                             \
+                (t*) alloca((sizeof(t)*_n_) ?: 1);                      \
```
Comment (keszybz) [inline src/basic/alloc-util.h:30]: Those new parentheses is not necessary... `*` has higher precedence than `?:`.

File: src/basic/alloc-util.h [hunk 2]
```diff
@@ -27,22 +27,22 @@ typedef void (*free_func_t)(void *p);
...
         })
 
 #define newa0(t, n)                                                     \
         ({                                                              \
                 size_t _n_ = n;                                         \
                 assert(!size_multiply_overflow(sizeof(t), _n_));        \
                 assert(sizeof(t)*_n_ <= ALLOCA_MAX);                    \
-                (t*) alloca0(sizeof(t)*_n_);                            \
+                (t*) alloca0((sizeof(t)*_n_) ?: 1);                     \
```
Comment (keszybz) [inline src/basic/alloc-util.h:38]: Same here.

**Review by keszybz — COMMENTED** [review]
LGTM. Only one cosmetic comment that can be fixed in a later PR.

**Pull request merged**